src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala

```diff
@@ -0,0 +1,351 @@
+    recheckBackOff.foreach(_.cancel())
+
+    super.postStop()
+
+    log.info("Stopped appTaskLaunchActor for {} version {}", app.id, app.version)
+  }
+
+  override def receive: Receive = waitForInitialDelay
+
+  private[this] def waitForInitialDelay: Receive = LoggingReceive.withLabel("waitingForInitialDelay") {
+    case RateLimiterActor.DelayUpdate(delayApp, delayUntil) if delayApp == app =>
+      stash()
+      unstashAll()
+      context.become(active)
+    case message: Any => stash()
+  }
+
+  private[this] def active: Receive = LoggingReceive.withLabel("active") {
+    Seq(
+      receiveStop,
+      receiveDelayUpdate,
+      receiveTaskLaunchNotification,
+      receiveTaskStatusUpdate,
+      receiveGetCurrentCount,
+      receiveAddCount,
+      receiveProcessOffers
+    ).reduce(_.orElse[Any, Unit](_))
+  }
+
+  private[this] def receiveWaitingForInFlight: Receive = LoggingReceive.withLabel("waitingForInFlight") {
+    case notification: TaskLaunchNotification =>
+      receiveTaskLaunchNotification(notification)
+      waitingForInFlight()
+
+    case "waitingForInFlight" => sender() ! "waitingForInFlight" // for testing
+  }
+
+  private[this] def receiveStop: Receive = {
+    case AppTaskLauncherActor.Stop => waitingForInFlight()
+  }
+
+  private[this] def waitingForInFlight(): Unit = {
+    if (inFlightTaskLaunches.isEmpty) {
+      context.stop(self)
+    }
+    else {
+      context.become(receiveWaitingForInFlight)
+    }
+  }
+
+  /**
+    * Receive rate limiter updates.
+    */
+  private[this] def receiveDelayUpdate: Receive = {
+    case RateLimiterActor.DelayUpdate(delayApp, delayUntil) if delayApp == app =>
+      if (backOffUntil.forall(_ < delayUntil)) {
+        import context.dispatcher
+        val now: Timestamp = clock.now()
+        recheckBackOff.foreach(_.cancel())
+        recheckBackOff = Some(
+          context.system.scheduler.scheduleOnce(now until delayUntil, self, RecheckIfBackOffUntilReached)
+        )
+      }
+
+      backOffUntil = Some(delayUntil)
+      OfferMatcherRegistration.manageOfferMatcherStatus()
+      log.debug("After delay update {}", status)
+
+    case RecheckIfBackOffUntilReached => OfferMatcherRegistration.manageOfferMatcherStatus()
+  }
+
+  private[this] def receiveTaskLaunchNotification: Receive = {
+    case TaskLaunchSourceDelegate.TaskLaunchRejected(taskInfo, reason) if inFlightTaskLaunches(taskInfo.getTaskId) =>
+      // This task is not yet known to mesos, so there will be no event that removes
+      // it automatically from the taskTracker.
+      taskTracker.terminated(app.id, taskInfo.getTaskId.getValue)
+      removeTask(taskInfo.getTaskId)
+      tasksToLaunch += 1
+      log.info(
+        "Task launch for '{}' was denied, reason '{}', rescheduling. {}",
+        taskInfo.getTaskId.getValue, reason, status)
+      OfferMatcherRegistration.manageOfferMatcherStatus()
+
+    case TaskLaunchSourceDelegate.TaskLaunchRejected(taskInfo, reason) => // ignore
```

Comment:
I think that the scheduler is fairly efficient. It's a hash wheel timer. Basically it has to maintain the same state that we have to maintain if we do not use it. And we already save something for every task anyway. So the algorithmic complexity should be the same.

So before "optimizing" it, we should do scale tests.